spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,201 @@
+
+  private[this] val writeConf = sqlContext.getWriteConf(keyspace, table, cluster)
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    data.rdd.saveToCassandra(keyspace, table, AllColumns, writeConf)(
+        new CassandraConnector(sqlContext.getCassandraConnConf(cluster)), SqlRowWriter.Factory)
+  }
+
+  def buildScan(): RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
+
+  protected[this] def getSchemaData(column: String, row: Row): (Any, NativeType) = {
+    val index = row.asInstanceOf[CassandraSQLRow].indexOf(column)
+    val columnDef = tableDef.columnByName.get(column).getOrElse(throw new IOException(s"Can't find $column"))
+    val dataType = ColumnDataType.catalystDataType(columnDef.columnType, nullable = true)
+    require(dataType.isPrimitive, s"${dataType.typeName} is not supported in filter.")
+
+    (row.get(index), dataType.asInstanceOf[NativeType])
+  }
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[TableScan]] */
+private[cassandra] class TableScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                               userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with TableScan {
+  override def buildScan(): RDD[Row] = super.buildScan()
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedScan]] */
+private[cassandra] class PrunedScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with PrunedScan {
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
+    val transformer = new RddFilterSelectPdTrf(requiredColumns, columnNameByLowercase, Seq.empty)
+    (transformer.maybeSelect apply baseRdd).asInstanceOf[RDD[Row]]
+  }
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedFilteredScan]] */
+private[cassandra] class PrunedFilteredScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                        userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with PrunedFilteredScan {
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val pushDown = new FilterPushDown(filters, tableDef)
+    val pushdownFilters = pushDown.toPushDown
+    val preservedFilters = pushDown.toPreserve
+
+    logInfo(s"pushdown filters: ${pushdownFilters.toString()}")
+
+    // a filter combining all other filters
+    def rowFilter(row: Row): Boolean = !preservedFilters.map(translateFilter(_)(row)).contains(false)
```

Comment:
`translateFilter` creates a function which translates a row. It is independent of a row so it can be computed once and then the result (which is a function) can be applied to each row, more or less something like this:
```scala
val translators = preservedFilters.map(translateFilter)
def rowFilter(row: Row): Boolean = translators.forAll(_(row))
```